### PR TITLE
grab cluster->lock when modifying exec_env->module_inst

### PR DESCRIPTION
**Pull request by yamt**
fixes: https://github.com/bytecodealliance/wasm-micro-runtime/issues/2680

also, when switching back to the original module_inst, propagate exception if any.

for a longer term, i think the exeception info should be moved from module_inst to exec_env.
cf. https://github.com/bytecodealliance/wasm-micro-runtime/issues/2512

### CODE REVIEW

File: core/iwasm/common/wasm_exec_env.c [hunk 1]
```diff
@@ -201,7 +201,51 @@ void
...
+void
+wasm_exec_env_restore_module_inst(
+    WASMExecEnv *exec_env, WASMModuleInstanceCommon *const module_inst_common)
+{
+    WASMModuleInstanceCommon *old_module_inst_common = exec_env->module_inst;
+    WASMModuleInstance *old_module_inst =
+        (WASMModuleInstance *)old_module_inst_common;
+    WASMModuleInstance *module_inst = (WASMModuleInstance *)module_inst_common;
+    char cur_exception[EXCEPTION_BUF_LEN];
+#if WASM_ENABLE_THREAD_MGR != 0
```
Comment (wenyongh) [inline core/iwasm/common/wasm_exec_env.c:222]: Had better add an empty line between L221 and L222.

**Review by wenyongh — COMMENTED** [review]
LGTM

File: core/iwasm/common/wasm_exec_env.c (hunk 1 above)
Comment (yamt) [reply core/iwasm/common/wasm_exec_env.c]: done.

**Review by xujuntwt95329 — COMMENTED** [review]
LGTM

**Review by wenyongh — APPROVED** [review]
LGTM

**Pull request merged**